### PR TITLE
fix(functions): resolve ~60 SonarCloud issues in Cloudflare functions

### PR DESCRIPTION
**Pull request by Wootehfook**
## Summary

Resolves all SonarCloud issues in `functions/**` (~60 issues: 5 Critical, ~15 Major, ~40 Minor).

### Cognitive complexity refactors — S3776 (Critical, ×5)
| Function | CC before | Approach |
|----------|-----------|----------|
| `watchlist-count` `scrapeWatchlistCount` | **39** | split into focused helpers: profile stats, metadata, content, JSON-LD, pagination, empty-state, poster-grid |
| `letterboxd/compare` `resolveMovie` | 25 | extracted search-title derivation, year-constrained + no-year lookup cascades, fallback builder |
| `test/movie-lookup` `findTmdbRowForMovie` | 26 | extracted `queryFirstRow` + candidate scoring/picking |
| `watchlist-count` `onRequestPost` | 21 | extracted cache-or-scrape resolution per username |
| `api/watchlist-count-updates` `validatePayload` | 21 | split into required/optional/username-format validators |
| `letterboxd/compare` `extractMoviesFromHtml` | 16 | extracted per-`<li>` scan step |
| `compare` batch closure | 17 | extracted `enhanceSingleMovie`, now reusing shared `parseGenresToNames` instead of duplicated genre parsing |

All refactors are behavior-preserving decompositions — same lookup order, same fallbacks, same logging.

### Super-linear regexes — S8786 (Major, ×8)
- Title/year splitting (`/^(.+?)\s+(\d{4})$/`) replaced with slice-based parsing — **verified byte-identical output on edge cases** (multi-year titles, multi-space, no-year, bare year)
- Whitespace-before-apostrophe collapse anchored on `(^|\S)` to stay linear
- Breadcrumb/page-heading counts now found in two linear steps (anchor match, then bounded 2KB window scan)
- friends-list year strip uses `exec().index` + `slice` instead of `\s*` regex

### Minor modernization (~40)
`Number.parseInt` with radix, `String#replaceAll`, `String.fromCodePoint`/`codePointAt`, `RegExp.exec`/`test` over `String#match`, optional chaining (incl. `_lib/common.js`), `Set#size`, merged duplicate import (S3863), unnecessary regex escapes (S6535), `stringifyError` helper to avoid `[object Object]` logs (S6551).

## Testing

- `tsc --noEmit` ✅
- `npm run test` — 117/117 ✅
- `npm run build` ✅
- Regex rewrites verified equivalent against edge-case corpus (see commit)

Part 3 of 4 SonarCloud cleanup PRs.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: functions/letterboxd/watchlist-count/index.ts [hunk 1]
```diff
@@ -30,219 +41,269 @@ async function rateLimit() {
...
+    return null;
+  }
+  try {
+    const jsonData = JSON.parse(jsonLdMatch[1]);
+    if (jsonData.numberOfItems || jsonData.totalCount) {
+      return jsonData.numberOfItems || jsonData.totalCount;
+    }
+  } catch {
+    // JSON parsing failed, continue with other methods
+  }
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/watchlist-count/index.ts:151]: Fixed in 03dd912: `findJsonLdCount` now coerces `numberOfItems`/`totalCount` via `Number()` and rejects non-finite/negative values, so a string (common in JSON-LD) or other type can no longer leak into the count path or the API response.

File: functions/letterboxd/watchlist-count/index.ts [hunk 2]
```diff
@@ -30,219 +41,269 @@ async function rateLimit() {
...
+  const directPatterns = [
+    /page\s+\d+\s+of\s+(\d+)/i,
+    /showing\s+[\d,-]+\s+of\s+(\d[\d,]*)/i,
+  ];
+  for (const pattern of directPatterns) {
+    const match = pattern.exec(html);
+    if (match) {
+      return { count: parseCountText(match[1]), estimated: false };
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/watchlist-count/index.ts]: Fixed in 03dd912: "Page X of Y" now estimates films as `pages * 72` (ITEMS_PER_PAGE) instead of returning the page count directly. The exact-total case ("showing … of N") is handled separately and stays exact.

File: functions/letterboxd/compare/index.ts [hunk 3]
```diff
@@ -267,7 +295,8 @@ async function enhanceWithTMDBData(
   const generateFallbackId = (t: string, y: number) => {
     let h = 0;
     const s = `${t.toLowerCase()}-${y}`;
-    for (let i = 0; i < s.length; i++) h = (h << 5) - h + s.charCodeAt(i);
+    for (let i = 0; i < s.length; i++)
+      h = (h << 5) - h + (s.codePointAt(i) ?? 0);
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/compare/index.ts]: Fixed in 03dd912: `generateFallbackId` now iterates with `for...of`, so each code point (including non-BMP surrogate pairs) contributes a single value instead of being double-counted as two UTF-16 units.

File: functions/compare/index.ts [hunk 4]
```diff
@@ -210,6 +211,75 @@ function findCommonMovies(
...
+function buildBasicMovie(movie: CommonMovie): Movie {
+  return {
+    id: Math.floor(Math.random() * 1000000),
+    title: movie.title,
+    year: movie.year,
+    letterboxdSlug: movie.slug,
+    friendCount: movie.friendCount, // Preserve friend information
+    friendList: movie.friendList, // Preserve friend information
+  };
+}
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/compare/index.ts:236]: Fixed in 03dd912: replaced the `Math.random()` fallback id with a deterministic title+year hash (`generateFallbackId`), so ids are stable across requests — no unstable React keys, no random collisions. This also clears the pseudorandom Sonar hotspot.

File: functions/test/movie-lookup.ts [hunk 5]
```diff
@@ -25,80 +25,100 @@ function normalizeTitle(t: string) {
...
+  return m ? Number.parseInt(m[1], 10) : 0;
+}
+
+// Run a lookup query and return the first row, or null when empty
+async function queryFirstRow(env: Env, sql: string, binds: unknown[]) {
+  const res = await env.MOVIES_DB.prepare(sql)
+    .bind(...binds)
+    .all();
+  return res.results && res.results.length > 0 ? res.results[0] : null;
+}
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/test/movie-lookup.ts]: Fixed in faf7875: `queryFirstRow` now uses `.first()` (each query is `LIMIT 1`), which returns the single row directly and avoids allocating the results array.

File: functions/letterboxd/watchlist-count/index.ts [hunk 6]
```diff
@@ -30,219 +41,279 @@ async function rateLimit() {
...
+  const pageOfMatch = /page\s+\d+\s+of\s+(\d+)/i.exec(html);
+  if (pageOfMatch) {
+    return {
+      count: Number.parseInt(pageOfMatch[1], 10) * ITEMS_PER_PAGE,
+      estimated: true,
+    };
+  }
+  // Estimate from last page link (same per-page assumption)
+  const lastPageMatch =
+    /<a[^>]+href="[^"]*page\/(\d+)"[^>]*>.*?(?:last|»|next)/i.exec(html);
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/watchlist-count/index.ts]: Fixed in faf7875: dropped `next` from the last-page alternation, so a next-page link is never misread as the final page (which would under-estimate the count). Only explicit `last`/`»` indicators are matched now.

File: functions/compare/index.ts [hunk 7]
```diff
@@ -1,7 +1,7 @@
 // AI Generated: GitHub Copilot - 2025-08-16
 // Letterboxd Watchlist Comparison API
 
-import { debugLog } from "../_lib/common";
+import { debugLog, parseGenresToNames } from "../_lib/common";
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/compare/index.ts]: Fixed in faf7875: updated the AI attribution header on functions/compare/index.ts to the required `// AI Generated: GitHub Copilot (<model-name>) - YYYY-MM-DD` format.

File: functions/compare/index.ts [hunk 8]
```diff
@@ -210,6 +211,85 @@ function findCommonMovies(
 // AI Generated: GitHub Copilot - 2025-08-29T12:00:00Z
 // Performance Optimization: Parallel Processing - Enhanced TMDB data processing with batch queries and concurrency control
 
+// Deterministic fallback id derived from title+year so that a movie without a
+// TMDB match keeps a stable id across requests (stable React keys, no random
+// collisions). Kept in the 900000+ range to avoid clashing with real TMDB ids.
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/compare/index.ts]: Fixed in 24447f3: reworded the comment to state this is best-effort — the 100k hash space can collide and the 900000+ offset only reduces, not eliminates, overlap with real TMDB ids. It remains a strict improvement over the previous `Math.random()` ids (stability), which was the point of the change.

File: functions/letterboxd/compare/index.ts [hunk 9]
```diff
@@ -94,10 +109,10 @@ function decodeHTMLEntities(text: string): string {
     (entity) => entityMap[entity] || entity
   );
   decoded = decoded.replace(/&#x([0-9A-Fa-f]+);/g, (_, hex) =>
-    String.fromCharCode(parseInt(hex, 16))
+    String.fromCodePoint(Number.parseInt(hex, 16))
   );
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/compare/index.ts]: Fixed in acb0ba5: added a `codePointToChar` guard (same pattern used in src/components/ResultsPage.tsx) that range-checks the value (integer, 0..0x10FFFF) before calling `String.fromCodePoint`, falling back to the original matched text otherwise. An oversized hex entity can no longer throw a RangeError.

File: functions/letterboxd/compare/index.ts [hunk 10]
```diff
@@ -94,10 +109,10 @@ function decodeHTMLEntities(text: string): string {
...
   decoded = decoded.replace(/&#(\d+);/g, (_, dec) =>
-    String.fromCharCode(parseInt(dec, 10))
+    String.fromCodePoint(Number.parseInt(dec, 10))
   );
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/compare/index.ts]: Fixed in acb0ba5: same `codePointToChar` guard applied to the decimal-entity branch.

File: functions/letterboxd/watchlist-count/index.ts [hunk 11]
```diff
@@ -5,20 +5,31 @@
  */
 
 import { debugLog } from "../../_lib/common";
-import type { D1DatabaseLike } from "../cache/index.js";
-import type { Env as CacheEnv } from "../cache/index.js";
+import type { D1DatabaseLike, Env as CacheEnv } from "../cache/index.js";
```
[the remark this answers is not in the document]
Comment (Wootehfook) [reply functions/letterboxd/watchlist-count/index.ts]: Fixed in acb0ba5: added `// AI Generated: GitHub Copilot (Claude Sonnet 5) - 2026-07-18` as the literal first line, and updated the block-comment copy to match the same format.
